src/simulators/superoperator/superoperator_state.hpp

```diff
@@ -0,0 +1,477 @@
+/**
+ * This code is part of Qiskit.
+ *
+ * (C) Copyright IBM 2018, 2019.
+ *
+ * This code is licensed under the Apache License, Version 2.0. You may
+ * obtain a copy of this license in the LICENSE.txt file in the root directory
+ * of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Any modifications or derivative works of this code must retain this
+ * copyright notice, and modified files need to carry a notice indicating
+ * that they have been altered from the originals.
+ */
+
+#ifndef _superoperator_state_hpp
+#define _superoperator_state_hpp
+
+#include <algorithm>
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+#include "framework/utils.hpp"
+#include "framework/json.hpp"
+#include "base/state.hpp"
+#include "superoperator.hpp"
+
+
+namespace AER {
+namespace QubitSuperoperator {
+
+// Allowed gates enum class
+enum class Gates {
+  u1, u2, u3, id, x, y, z, h, s, sdg, t, tdg, // single qubit
+  cx, cz, swap, // two qubit
+  ccx // three qubit
+};
+
+// Allowed snapshots enum class
+enum class Snapshots {superop};
+
+//=========================================================================
+// QubitUnitary State subclass
+//=========================================================================
+
+template <class data_t = QV::Superoperator<double>>
+class State : public Base::State<data_t> {
+public:
+  using BaseState = Base::State<data_t>;
+
+  State() = default;
+  virtual ~State() = default;
+
+  //-----------------------------------------------------------------------
+  // Base class overrides
+  //-----------------------------------------------------------------------
+
+  // Return the string name of the State class
+  virtual std::string name() const override {return "superoperator";}
+
+  // Return the set of qobj instruction types supported by the State
+  virtual Operations::OpSet::optypeset_t allowed_ops() const override {
+    return Operations::OpSet::optypeset_t({
+      Operations::OpType::gate,
+      Operations::OpType::reset,
+      Operations::OpType::snapshot,
+      Operations::OpType::barrier,
+      Operations::OpType::matrix,
+      Operations::OpType::kraus,
+      Operations::OpType::superop
+    });
+  }
+
+  // Return the set of qobj gate instruction names supported by the State
+  virtual stringset_t allowed_gates() const override {
+    return {"U", "CX", "u1", "u2", "u3", "cx", "cz", "swap",
+            "id", "x", "y", "z", "h", "s", "sdg", "t", "tdg", "ccx"};
+  }
+
+  // Return the set of qobj snapshot types supported by the State
+  virtual stringset_t allowed_snapshots() const override {
```

Comment:
In our TODO list, we have to decople Qobj as much as possible.. just in case we change the format in the future (Protobuffer... is coming...)